### PR TITLE
Attachments transform

### PR DESCRIPTION
**Pull request by simonexmachina**
I think that it's important to provide a single attachment transform because it's simpler and easier to understand. This PR is a suggestion, and I haven't added tests, but you can see that the single attachment use case can be very easily layered over the underlying multiple attachment transform.

### CODE REVIEW

**Comment by simonexmachina** [comment]
I just realised that this includes the `Ember.get` code as well. Happy to remove if you prefer.

**Comment by stevebest** [comment]
I like the idea of having a singular 'attachment'-type attributes, but I don't quite like the implementation. Imagine that a document somehow has several attachments. In a form that you suggest, this transform `pop`s basically random one and forgets about all others. This is only usable if you only declare exactly one `DS.attr('attachment')` and make sure that you don't have any other writers that could add more than one attachment to your doc.

What I originally thought of was something like that:

``` js
const User = DS.Model.extend({
  // . . .
  picture:  DS.attr('attachment'),
  ringtone: DS.attr('attachment'),
  files:    DS.attr('attachments'),
});
```

which translates to

``` js
{
  "_attachments": {
    "picture": { /* . . . */ },
    "hello.txt": { /* . . . */ }
  }
}
```

but this is obviously not trivial to implement. Ember Data [Transforms](http://emberjs.com/api/data/classes/DS.Transform.html) can only look at one attribute value at a time, so I'm not sure how exactly you could put all the values into a resulting `_attachments` object. Seems like that would require adding a specialized [Serializer](http://emberjs.com/api/data/classes/DS.Serializer.html).

In addition, I thought about maybe having an `Attachment` object implementing the functionality of holding the `name`, `data`, and other attributes.

**Comment by simonexmachina** [comment]
Hi @stevebest I've added a commit which adds functionality to the serializer so that the attachments are passed to and from relational-couch correctly. This works for single and multiple attachments.

The single `attachment` only ever sees the one attachment, so there's no randomness in which field it yields.

There's no test coverage for the serializer changes, but it does appear to be working. I'm afraid I won't be able to find the time to add tests for this because I'm under the pump on this project and have other open source projects that are already waiting for my attention.

Let me know how you'd like to proceed with this. Happy to jump on a call or whatever to work out the best way to proceed. Cheers!

**Comment by stevebest** [comment]
Hey, Simon! Overall this looks more like it. Great job!

I believe @nolanlawson will rightfully insist on writing tests for this PR. I'd be glad to review and finish it up, but I've got an even better excuse to slack off: I'm getting married next week! :bride_with_veil: :ring:

---

Anyhow, there are some corner cases I'd like to address:
- Having more than one `attr('attachment')` is certainly fine, but what about multiple `attr('attachments')`? We probably should detect it and fail early, like models without `rev`.
- Should `attr('attachments')` include _all_ attachments, even if we have one or several specific `attr('attachment')`s? Let's say we have:

``` js
  picture: DS.attr('attachment'),   // singular
  files:   DS.attr('attachments'),  // plural
```

What should happen if we `myModel.get('files').addObject({ name: 'picture' })`? What about `myModel.set('picture', { ... })`? Maybe add a method to a model, like `model.attach({ name: 'picture' })`?
- What if we only specify singular `attr('attachment')`s on a model, and a doc happens to have an attachment we didn't account for? Say a doc has `{_attachments: {foo: ..., bar: ...}}` and we only declared `foo: DS.attr('attachment')`. If we load and save this model, `bar` attachment will be lost, and it is sad. (Although, same goes for any other attribute, attachment or not.)
- Integration tests against actual PouchDB and CouchDB would be awesome. There seems to be a little discrepancy between them when it comes to attachments.

**Comment by nolanlawson** [comment]
Hey, bit off topic, but can I ask if there was an issue in `nolanlawson/ember-pouch` that necessitated a fork? I haven't had much time to keep `ember-pouch` updated myself, but if this fork is better maintained, then I'm happy to hand over commit bits and npm publish access.

As for this PR, yes, I am usually pretty severe about having tests to go along with PRs (it's far too easy to unintentionally break things otherwise).

**Comment by simonexmachina** [comment]
Hey @nolanlawson, no issues with ember-pouch just wanted to add attachments support. My fork was made to submit improvements to this PR.

I've given it some thought and I'd be open to becoming a maintainer for this project. I'm the author and maintainer of [ember-notify](https://github.com/aexmachina/ember-notify), [ember-cli-sass](https://github.com/aexmachina/ember-cli-sass) and [ember-devtools](https://github.com/aexmachina/ember-devtools). My team at [jtribe](http://jtribe.com.au/) is actively using ember-couch on our current project and it's providing an architecture that we'll probably continue using in the future, so I think it makes sense for us to continue contributing. We may be able to help move things forward, either here or on our fork - do you have a big picture of how you'd like to see the project progressing?

**Comment by simonexmachina** [comment]
Hey @stevebest, see responses below:

> I believe @nolanlawson will rightfully insist on writing tests for this PR. I'd be glad to review and finish it up, but I've got an even better excuse to slack off: I'm getting married next week! 👰 💍

Yep, I think that it's important to add tests. Happy to do so, but it might take me a while - I've got a few other open source things on my list already.

Congratulations on the wedding! If you're taking some time off then maybe I should submit these changes in another PR from my fork?

> multiple `attr('attachments')`

This is supported. The serializer places the correct set of `_attachments` entries from the response into each attribute, based on the file names. There is a potential collision with mulitiple files with the same file name, but that's part of the underlying CouchDB attachments mechanism so a solution is I think outside the scope of this PR.

> Should `attr('attachments')` include all attachments

See response above - each attribute gets the correct set of attachments.

> What if we only specify singular `attr('attachment')`s on a model, and a doc happens to have an attachment we didn't account for?

Only the file(s) that have been created for an attribute are accessible through that attribute.

> Say a doc has {_attachments: {foo: ..., bar: ...}} and we only declared foo: DS.attr('attachment'). If we load and save this model, bar attachment will be lost, and it is sad. (Although, same goes for any other attribute, attachment or not.)

Yes this is true, but I think this correctly retainins the semantics of the rest of ember-data. The alternative is that the `attachments` serializer exposes all the `_attachments`, but then it would only make sense to have a single `attr('attachments')` on a given model. I'm interested to know your feeling on this, but my feeling is that it's correct the way it is now.

> Integration tests against actual PouchDB and CouchDB would be awesome. There seems to be a little discrepancy between them when it comes to attachments.

I think this is outside the scope of this package. PouchDB is responsible for correctly transferring attachments to CouchDB, not ember-couch.

**Comment by nolanlawson** [comment]
@aexmachina Cool, you are now a commiter. Let me know your npm username and I'll make you an owner, too. @stevebest you are now a committer as well.

I think it's more important to have passionate people who are actually using this software on board as maintainers. TBQH ember-pouch was a weekend project for me that quickly grew beyond its original vision and was nurtured and enhanced by folks who were capable of things I couldn't even understand. Around the same time, I lost interest in ember-pouch and moved on to other things. Since then I've barely tinkered with Ember and wouldn't even know how to start contributing to the project.

If it makes you uncomfortable that it's under `github.com/nolanlawson` I would be more than happy to create an `ember-pouch` group and put it there; it feels silly to take credit for something that isn't really mine anymore. Otherwise please feel free to keep contributing to `nolanlawson/ember-pouch` and take it in the direction you see fit. :)

File: addon/serializers/pouch.js [hunk 1]
```diff
@@ -15,5 +21,41 @@ export default DS.RESTSerializer.extend({
...
+      // if provided, use the mapping provided by `attrs` in the serializer
+      var payloadKey = this._getMappedKey(key, snapshot.type);
+      if (payloadKey === key && this.keyForAttribute) {
+        payloadKey = this.keyForAttribute(key, 'serialize');
+      }
+      // assign any attachments to the attachments property, so that relational-pouch
+      // will put these in the special CouchDB _attachments property
+      // this will conflict with any 'attachments' attr in the model
+      // suggest that #toRawDoc in relational-pouch should allow _attachments to be specified
+      json['attachments'] = Object.assign({}, json['attachments'], json[payloadKey]);
```
Comment (simonexmachina) [inline addon/serializers/pouch.js]: @nolanlawson if the user has an `attr` called 'attachments' in their model then it will be overwritten here. Can we change #toRawDoc in relational-pouch to allow attachments to be provided using an `_attachments` property?

**Comment by simonexmachina** [comment]
I've started work on some integration tests for the attachments support in the serializer.

**Comment by simonexmachina** [comment]
@stevebest I've added an integration test for the serializer now. It feels strange pushing to this fork rather than to ember-pouch. Do you want me to open a new PR that contains both of our changes?

**Comment by simonexmachina** [comment]
@stevebest I'm going to open another PR that contains these changes on ember-pouch.

**Comment by simonexmachina** [comment]
@nolanlawson, what about moving ember-pouch to pouchdb/ember-pouch?

**Comment by nolanlawson** [comment]
I think the other Pouch contributors might feel uncomfortable; might make it look like we are advocating one framework over another. I only proposed moving it out of `nolanlawson` if it makes you uncomfortable; otherwise it's fine with me to keep it where it is. :)

**Comment by simonexmachina** [comment]
Okay no worries.​
